src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,673 @@
+		return NULL;
+	}
+
+	while (list_length(state->truncations) > 0) {
+		// Get the first item from the list.
+		PossibleTruncation *truncation = linitial(state->truncations);
+		// ... and then remove it from the list.
+		state->truncations = list_delete_nth_cell(state->truncations, 0);
+
+		Node *node = truncation->node;
+		enum TruncationAttr attr = truncation->attr;
+
+		if (truncation->length <= 3) {
+			// If "truncating" would make it longer, refuse to truncate.
+		} else if (IsA(node, SelectStmt) && attr == TRUNCATION_TARGET_LIST) {
+			SelectStmt *stmt = castNode(SelectStmt, node);
+			stmt->targetList = list_make1(dummy_target());
+			printf("Select/targetList\n");
+		}
+		else if (IsA(node, SelectStmt) && attr == TRUNCATION_WHERE_CLAUSE) {
+			SelectStmt *stmt = castNode(SelectStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("Select/whereClause\n");
+		}
+		else if (IsA(node, SelectStmt) && attr == TRUNCATION_VALUES_LISTS) {
+			SelectStmt *stmt = castNode(SelectStmt, node);
+			stmt->valuesLists = list_make1(dummy_column());
+			printf("Select/valuesLists\n");
+		}
+		else if (IsA(node, UpdateStmt) && attr == TRUNCATION_TARGET_LIST) {
+			UpdateStmt *stmt = castNode(UpdateStmt, node);
+			stmt->targetList = list_make1(dummy_target());
+			printf("UpdateStmt/targetList\n");
+		}
+		else if (IsA(node, UpdateStmt) && attr == TRUNCATION_WHERE_CLAUSE) {
+			UpdateStmt *stmt = castNode(UpdateStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("UpdateStmt/whereClause\n");
+		}
+		else if (IsA(node, DeleteStmt) && attr == TRUNCATION_WHERE_CLAUSE) {
+			DeleteStmt *stmt = castNode(DeleteStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("Delete/whereClause\n");
+		}
+		else if (IsA(node, CopyStmt) && attr == TRUNCATION_WHERE_CLAUSE) {
+			CopyStmt *stmt = castNode(CopyStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("Copy/whereClause\n");
+		}
+		else if (IsA(node, InsertStmt) && attr == TRUNCATION_COLS) {
+			InsertStmt *stmt = castNode(InsertStmt, node);
+			stmt->cols = list_make1(dummy_target());
+			printf("Insert/cols\n");
+		}
+		else if (IsA(node, IndexStmt) && attr == TRUNCATION_WHERE_CLAUSE) {
+			IndexStmt *stmt = castNode(IndexStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("Index/whereClause\n");
+		}
+		else if (IsA(node, RuleStmt) && attr == TRUNCATION_WHERE_CLAUSE) {
+			RuleStmt *stmt = castNode(RuleStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("Rule/whereClause\n");
+		}
+		else if (IsA(node, CommonTableExpr) && attr == TRUNCATION_CTE_QUERY) {
+			// Equivalent to: https://github.com/pganalyze/pg_query.rs/blob/66eb7becea1a40e315fee3f90197a35a89d20c25/src/truncate.rs#L230-L237
+			CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+			Node *old = stmt->ctequery;
+			stmt->ctequery = dummy_select(NULL, (Node *) dummy_column(), NULL);
+			if (old) {
```

Comment:
To capture my understanding: This logic (that came from Rust originally, and is not present in the Ruby implementation) helps us not revisit the statement contained (e.g. `SelectStmt`), because we already removed the whole CTE anyway. Because it is valid to update a removed node (since we do not free it, we just remove it from the tree), this should not affect correctness, but rather it lets us potentially skip an unnecessary deparse.

However, I think because we go depth first, this may never be utilized, so its potentially worth dropping here.

Reply:
Also, if we remove this, we could replace the iteration logic with a foreach, since we don't actually have to modify the list as we iterate over it.